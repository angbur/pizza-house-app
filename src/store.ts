import { configureStore } from '@reduxjs/toolkit';
import orderReducer from 'components/Pages/Order/orderSlice';
import { pizzaApi } from 'services/pizza';

export const store = configureStore({
  reducer: {
    order: orderReducer,
    [pizzaApi.reducerPath]: pizzaApi.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(pizzaApi.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
