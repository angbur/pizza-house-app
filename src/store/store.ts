import { configureStore, Store } from '@reduxjs/toolkit';
import orderReducer from 'components/Pages/Order/orderSlice';
import dialogReducer from 'components/Organisms/Dialog/dialogSlice';
import { pizzaApi } from 'services/pizza';
import orderMiddleware from './middleware/orderMiddleware';

export const store: Store = configureStore({
  reducer: {
    order: orderReducer,
    dialog: dialogReducer,
    [pizzaApi.reducerPath]: pizzaApi.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(pizzaApi.middleware, orderMiddleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
