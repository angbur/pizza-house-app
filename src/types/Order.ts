import type { Pizza } from 'types/Pizza';

export type OrderItem = {
  _id: string;
  item: Omit<Pizza, 'categories' | '_id'>;
  quantity: number;
};
