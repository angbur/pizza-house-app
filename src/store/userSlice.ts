import { createSlice } from '@reduxjs/toolkit';
import { RootState } from './store';
import { toast } from 'react-toastify';
import { userApi } from 'services/user';

type User = {
  loggedIn: boolean;
  status: string | undefined;
  error: string | undefined;
  userId: string | undefined;
  login: string | undefined;
  token: string | undefined;
};

const initialState: User = {
  loggedIn: false,
  status: 'idle',
  error: undefined,
  userId: undefined,
  login: undefined,
  token: undefined,
};

const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    logout: (state) => {
      state.loggedIn = false;
      state.userId = undefined;
      state.login = undefined;
      state.token = undefined;
      toast.success('Logged out!');
    },
  },
  extraReducers: (builder) => {
    builder.addMatcher(userApi.endpoints.login.matchFulfilled, (state, { payload }) => {
      state.loggedIn = true;
      state.token = payload.token;
      state.userId = payload.userId;
      state.login = payload.login;
    });
  },
});

export const { logout } = userSlice.actions;

export const selectName = (state: RootState) => state.user.user;

export const isLoggedIn = (state: RootState) => state.user.loggedIn;

export const selectResponseStatus = (state: RootState) => state.user.status;

export const selectResponseError = (state: RootState) => state.user.error;

export const selectLoggedInUser = (state: RootState) => state.user.user;

export default userSlice.reducer;
