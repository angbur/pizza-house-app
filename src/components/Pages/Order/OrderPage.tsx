import Button from 'components/Atoms/Button/Button';
import Typography from 'components/Atoms/Typography/Typography';
import { openDialog, FormType } from 'components/Organisms/Dialog/dialogSlice';
import { useAppDispatch, useAppSelector } from 'hooks';
import { useNavigate } from 'react-router-dom';
import { selectIsLoggedIn } from 'store/userSlice';
import styled from 'styled-components';
import OrderList from './components/OrderList/OrderList';
import { removeAllOrder, selectListOfOrderItems } from './orderSlice';

const Container = styled.section`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  text-align: center;
  width: 100%;
  & > h2 {
    margin: 2rem 0;
  }
`;

const StyledDiv = styled.div`
  display: flex;
  gap: 3rem;
  margin: 2rem;
`;

const OrderPage = () => {
  const navigate = useNavigate();
  const dispatch = useAppDispatch();
  const isLoggedIn = useAppSelector(selectIsLoggedIn);
  const orderList = useAppSelector(selectListOfOrderItems);

  const handleRemoveAll = () => {
    dispatch(removeAllOrder());
  };

  const handleNavigate = () => {
    if (!isLoggedIn) dispatch(openDialog(FormType.login));
    navigate('/order/form');
  };
  return (
    <Container>
      <Typography variant='section-title' color='light'>
        My Order
      </Typography>
      <OrderList />
      {orderList.length !== 0 ? (
        <StyledDiv>
          <Button variant='button-text-light' onClick={handleRemoveAll}>
            Remove all
          </Button>
          <Button variant='primary-light' onClick={handleNavigate}>
            Go to order form
          </Button>
        </StyledDiv>
      ) : null}
    </Container>
  );
};

export default OrderPage;
