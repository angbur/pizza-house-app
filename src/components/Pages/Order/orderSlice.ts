import { OrderItem } from 'types/Order';
import { createSlice, nanoid, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from 'store/store';
import { toast } from 'react-toastify';
import { logout } from 'store/user/userSlice';

type RequireOnly<T, P extends keyof T> = Pick<T, P> & Partial<Omit<T, P>>;

export type DraftItem = RequireOnly<OrderItem, 'item' | 'quantity'>;

export const createItem = (draftItem: DraftItem): OrderItem => {
  return { _id: nanoid(), ...draftItem };
};

type OrderState = {
  entities: OrderItem[];
};

const getInitialState = () => {
  let result;
  result = localStorage.getItem('order');
  if (typeof result === 'string') {
    result = JSON.parse(result);
  }
  return result;
};

const initialState: OrderState = {
  entities: getInitialState()?.entities || [],
};

const orderSlice = createSlice({
  name: 'order',
  initialState,
  reducers: {
    addToOrder: (state, action: PayloadAction<DraftItem>) => {
      const existingIndex = state.entities.findIndex(
        (el) => el.item.name === action.payload.item.name,
      );
      if (existingIndex === -1) {
        const item = createItem(action.payload);
        state.entities.unshift(item);
        toast.success('Added to order!');
      } else {
        state.entities[existingIndex].quantity += action.payload.quantity;
        toast.success('Order updated');
      }
    },
    removeFromOrder: (state, action: PayloadAction<OrderItem['_id']>) => {
      const index = state.entities.findIndex((item) => item._id === action.payload);
      state.entities.splice(index, 1);
    },
    updateOrderItem: (state, action: PayloadAction<OrderItem>) => {
      const index = state.entities.findIndex((item) => item._id === action.payload._id);
      state.entities[index] = action.payload;
    },
    removeAllOrder: (state) => {
      state.entities = [];
      toast.success('The order has been deleted!');
    },
  },
  extraReducers: (builder) => {
    builder.addCase(logout, (state) => {
      state.entities = [];
    });
  },
});

export const { addToOrder, removeFromOrder, updateOrderItem, removeAllOrder } = orderSlice.actions;

export default orderSlice.reducer;

export const selectListOfOrderItems = (state: RootState) => state.order.entities;
export const selectSumOfOrder = (state: RootState) =>
  state.order.entities.reduce(
    (prev: number, curr: OrderItem) => prev + curr.item.price * curr.quantity,
    0,
  );
