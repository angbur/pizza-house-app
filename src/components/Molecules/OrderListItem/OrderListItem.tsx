import Button from 'components/Atoms/Button/Button';
import Icon from 'components/Atoms/Icon/Icon';
import PriceTag from 'components/Atoms/PriceTag/PriceTag';
import QuantityChangeButton from 'components/Atoms/QuantityChangeButton/QuantityChangeButton';
import Typography from 'components/Atoms/Typography/Typography';
import { removeFromOrder, updateOrderItem } from 'components/Pages/Order/orderSlice';
import { ThemeContext } from 'components/Theme/ThemeContext';
import { useAppDispatch } from 'hooks';
import { useContext } from 'react';
import styled from 'styled-components';
import { OrderItem } from 'types/Order';
import ButtonWithInputMobile from '../ButtonWithInputMobile/ButtonWithInputMobile';

const StyledListItem = styled.li`
  border-radius: 8px;
  border: 2px solid ${(props) => props.theme.palette.light};
  display: flex;
  align-items: center;
  gap: 0.625rem;
  padding: 1rem 2rem;
  width: 800px;
  & > .quantity-container {
    position: relative;
    top: 4px;
  }
  & .quantity-button-mobile {
    display: none;
  }
  & > img {
    margin-left: 2rem;
  }
  @media (max-width: 950px) {
    width: auto;
    margin: 0 2rem;
  }
  @media (max-width: 440px) {
    margin: 0;
  }
  @media (max-width: 640px) {
    & img.pizza-img {
      display: none;
    }
  }
  @media (max-width: 545px) {
    & img.pizza-img,
    .quantity-container,
    .quantity-button,
    .icon-button {
      display: none;
    }
    padding: 1rem;
    display: flex;
    flex-direction: column-reverse;
    gap: 1rem;
    & .quantity-button-mobile {
      display: block;
    }
    & .quantity-button {
      display: none;
    }
  }
`;

const DescriptionDiv = styled.div`
  display: flex;
  flex-direction: column;
  align-items: start;
  gap: 0.325rem;
  padding: 0 1rem;
  flex: 50%;
  text-align: left;
  @media (max-width: 545px) {
    flex: 100%;
    padding: 0;
    margin-bottom: 1rem;
  }
`;

const QuantityDiv = styled.div`
  flex: 20%;
`;

const PriceDiv = styled.div`
  flex: 10%;
  display: flex;
  justify-content: center;
  align-items: center;
    & > .price-tag {
      position: relative;
      top: 4px;
    }
    @media (max-width: 545px) {
      & > p {
        margin-right: 1rem;
      }
      width: 100%;
      justify-content: flex-end;
    }
  }
`;

export type OrderListItemProps = {
  item: OrderItem;
  className?: string;
};

const OrderListItem = ({ item }: OrderListItemProps) => {
  const theme = useContext(ThemeContext);
  const dispatch = useAppDispatch();

  const handleOnIncrement = () => {
    const newItem = { ...item };
    newItem.quantity += 1;
    dispatch(updateOrderItem(newItem));
  };
  const handleOnDecrement = () => {
    const newItem = { ...item };
    if (item.quantity != 1) {
      newItem.quantity -= 1;
      dispatch(updateOrderItem(newItem));
    }
    if (item.quantity === 1) {
      dispatch(removeFromOrder(item._id));
    }
  };

  const handleRemove = () => dispatch(removeFromOrder(item._id));

  return (
    <StyledListItem theme={theme}>
      <ButtonWithInputMobile
        value={item.quantity}
        onClick={handleRemove}
        onIncrement={handleOnIncrement}
        onDecrement={handleOnDecrement}
        withRemove={true}
      />
      <QuantityDiv className={'quantity-container'}>
        <Typography variant='price-tag' color='light'>
          {item.quantity} X
        </Typography>
      </QuantityDiv>
      <QuantityChangeButton onIncrement={handleOnIncrement} onDecrement={handleOnDecrement} />
      <img
        src={require(`assets/image/pizza-${item.item._id}.svg`)}
        height={'75px'}
        className={'pizza-img'}
        alt={`Picture of ${item.item.name} pizza`}
      />
      <DescriptionDiv className={'description'}>
        <Typography variant='card-title' style={{ ...{ textTransform: 'uppercase' } }}>
          {item.item.name}
        </Typography>
        <Typography variant='card-subtitle'>
          {item.item.ingredients.map((i, id) =>
            id != item.item.ingredients.length - 1 ? (
              <span key={`ingredients-${i}`}> {i}, </span>
            ) : (
              <span key={`ingredients-${i}`}> {i}</span>
            ),
          )}
        </Typography>
      </DescriptionDiv>
      <PriceDiv>
        <PriceTag price={item.item.price * item.quantity} size='md' />
      </PriceDiv>
      <Button variant={'icon-button'} onClick={handleRemove}>
        {Icon({ name: 'trash', size: 14, isActive: false })}
      </Button>
    </StyledListItem>
  );
};

export default OrderListItem;
