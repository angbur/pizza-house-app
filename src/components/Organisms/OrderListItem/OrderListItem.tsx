import Button from 'components/Atoms/Button/Button';
import Icon from 'components/Atoms/Icon/Icon';
import PriceTag from 'components/Atoms/PriceTag/PriceTag';
import QuantityChangeButton from 'components/Atoms/QuantityChangeButton/QuantityChangeButton';
import Typography from 'components/Atoms/Typography/Typography';
import { removeFromOrder, updateOrderItem } from 'components/Pages/Order/orderSlice';
import { ThemeContext } from 'components/Theme/ThemeContext';
import { useAppDispatch } from 'hooks';
import { useContext } from 'react';
import styled from 'styled-components';
import { OrderItem } from 'types/Order';

const StyledListItem = styled.li`
  border-radius: 8px;
  border: 2px solid ${(props) => props.theme.palette.light};
  display: flex;
  align-items: center;
  gap: 0.625rem;
  padding: 1rem 2rem;
  max-width: 800px;
  & > .quantity-container {
    position: relative;
    top: 4px;
  }
  & > img {
    margin-left: 2rem;
  }
`;

const DescriptionDiv = styled.div`
  display: flex;
  flex-direction: column;
  align-items: start;
  gap: 0.325rem;
  padding: 0 1rem;
  flex: 50%;
  text-align: left;
`;

const QuantityDiv = styled.div`
  flex: 20%;
`;

const PriceDiv = styled.div`
  flex: 10%;
  display: flex;
  justify-content: center;
  align-items: center;
  & > .price-tag {
    position: relative;
    top: 4px;
  }
`;

export type OrderListItemProps = {
  item: OrderItem;
};

const OrderListItem = ({ item }: OrderListItemProps) => {
  const theme = useContext(ThemeContext);
  const dispatch = useAppDispatch();

  const handleOnIncrement = () => {
    const newItem = { ...item };
    newItem.quantity += 1;
    dispatch(updateOrderItem(newItem));
  };
  const handleOnDecrement = () => {
    const newItem = { ...item };
    if (item.quantity != 1) {
      newItem.quantity = +1;
      dispatch(updateOrderItem(newItem));
    }
    if (item.quantity === 1) {
      dispatch(removeFromOrder(item._id));
    }
  };

  const handleRemove = () => dispatch(removeFromOrder(item._id));

  return (
    <StyledListItem theme={theme}>
      <QuantityDiv className={'quantity-container'}>
        <Typography variant='price-tag' color='light'>
          {item.quantity} X
        </Typography>
      </QuantityDiv>
      <QuantityChangeButton onIncrement={handleOnIncrement} onDecrement={handleOnDecrement} />
      <img
        src={require(`assets/image/pizza-${item.item._id}.svg`)}
        height={'75px'}
        alt={`Picture of ${item.item.name} pizza`}
      />
      <DescriptionDiv>
        <Typography variant='card-title' style={{ ...{ textTransform: 'uppercase' } }}>
          {item.item.name}
        </Typography>
        <Typography variant='card-subtitle'>
          {item.item.ingredients.map((i, id) =>
            id != item.item.ingredients.length - 1 ? (
              <span key={`ingredients-${i}`}> {i}, </span>
            ) : (
              <span key={`ingredients-${i}`}> {i}</span>
            ),
          )}
        </Typography>
      </DescriptionDiv>
      <PriceDiv>
        <PriceTag price={item.item.price * item.quantity} size='md' />
      </PriceDiv>
      <Button variant={'icon-button'} onClick={handleRemove}>
        {Icon({ name: 'trash', size: 14, isActive: false })}
      </Button>
    </StyledListItem>
  );
};

export default OrderListItem;
