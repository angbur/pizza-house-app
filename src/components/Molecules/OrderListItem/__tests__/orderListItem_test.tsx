import renderer from 'react-test-renderer';
import ThemeProvider from 'components/Theme/ThemeContext';
import { store } from 'store/store';
import { Provider } from 'react-redux';
import { OrderItem } from 'types/Order';
import OrderListItem from '../OrderListItem';

type WrapperProps = {
  item: OrderItem;
};

const Wrapper = ({ item }: WrapperProps) => {
  return (
    <ThemeProvider>
      <Provider store={store}>
        <OrderListItem item={item} />
      </Provider>
    </ThemeProvider>
  );
};

const orderItem: OrderItem = {
  _id: 'awesome_id',
  item: {
    _id: '63b2d3b592b7f0f1c7c567b8',
    name: 'Pinza Mare',
    ingredients: ['shrimps', 'red fish', 'olives', 'white sauce', 'oregano', 'basil'],
    price: 26,
  },
  quantity: 1,
};

describe('Order List item', () => {
  it('should renders order list item correctly', () => {
    const tree = renderer.create(<Wrapper item={orderItem} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
