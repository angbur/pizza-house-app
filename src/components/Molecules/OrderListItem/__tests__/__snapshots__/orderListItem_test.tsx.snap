// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order List item should renders order list item correctly 1`] = `
<li
  className="sc-csuSiG hBwBHz"
>
  <div
    className="sc-ipEyDJ caIvtJ quantity-button-mobile"
  >
    <button
      className="sc-bcXHqe sc-gswNZR feNVH gHullA primary-light btn-md button"
      onClick={[Function]}
      style={
        Object {
          "border": "none",
          "borderRadius": "0",
          "fontSize": "0.9500000000000001rem",
          "maxHeight": "80px",
          "maxWidth": "80px",
          "padding": "0.5850000000000001rem 1.125rem",
        }
      }
    >
      <img
        height={20}
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "plus.svg",
          }
        }
      />
    </button>
    <input
      className="sc-kDvujY dYXGJW"
      style={
        Object {
          "border": "none",
          "borderRadius": "0",
          "maxHeight": "80px",
          "maxWidth": "80px",
        }
      }
      type="number"
      value={1}
    />
    <button
      className="sc-bcXHqe sc-gswNZR feNVH gHullA primary-light btn-md button"
      onClick={[Function]}
      style={
        Object {
          "border": "none",
          "borderRadius": "0",
          "fontSize": "0.9500000000000001rem",
          "maxHeight": "80px",
          "maxWidth": "80px",
          "padding": "0.5850000000000001rem 1.125rem",
        }
      }
    >
      <img
        height={20}
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "minus.svg",
          }
        }
      />
    </button>
    <button
      className="sc-bcXHqe sc-gswNZR feNVH gHullA primary-light btn-md button"
      onClick={[Function]}
      style={
        Object {
          "border": "none",
          "borderLeft": "1px solid white",
          "borderRadius": "0",
          "fontSize": "0.9500000000000001rem",
          "maxHeight": "80px",
          "maxWidth": "80px",
          "padding": "0.5850000000000001rem 1.125rem",
        }
      }
    >
      <img
        height={20}
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "trash.svg",
          }
        }
      />
    </button>
  </div>
  <div
    className="sc-bqWxrE gbPZDU quantity-container"
  >
    <p
      className="price-tag"
      style={
        Object {
          "color": "#FCFBFB",
          "fontFamily": "Klepon Ijo",
          "fontSize": "3rem",
          "letterSpacing": "0",
        }
      }
    >
      1
       X
    </p>
  </div>
  <div
    className="sc-jrcTuL ggyCiw quantity-button"
  >
    <button
      onClick={[Function]}
    >
      <img
        height={14}
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "arrow-up-active.svg",
          }
        }
      />
    </button>
    <button
      onClick={[Function]}
    >
      <img
        height={14}
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "arrow-down-active.svg",
          }
        }
      />
    </button>
  </div>
  <img
    alt="Picture of Pinza Mare pizza"
    className="pizza-img"
    height="75px"
    src={
      Object {
        "ReactComponent": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "__esModule": true,
        "default": "pizza-63b2d3b592b7f0f1c7c567b8.svg",
      }
    }
  />
  <div
    className="sc-eDWCr geClNj description"
  >
    <h3
      className="card-title"
      style={
        Object {
          "color": "none",
          "fontFamily": "Klepon Ijo",
          "fontSize": "2rem",
          "fontWeight": "400",
          "letterSpacing": "2px",
          "lineHeight": "1.2",
          "textTransform": "uppercase",
        }
      }
    >
      Pinza Mare
    </h3>
    <h4
      className="card-subtitle"
      style={
        Object {
          "color": "none",
          "fontFamily": "Source Sans Pro",
          "fontSize": "0.9rem",
          "fontWeight": "400",
          "lineHeight": "1.6",
        }
      }
    >
      <span>
         
        shrimps
        , 
      </span>
      <span>
         
        red fish
        , 
      </span>
      <span>
         
        olives
        , 
      </span>
      <span>
         
        white sauce
        , 
      </span>
      <span>
         
        oregano
        , 
      </span>
      <span>
         
        basil
      </span>
    </h4>
  </div>
  <div
    className="sc-ksBlkl hWYqYS"
  >
    <p
      className="price-tag"
      style={
        Object {
          "color": "#EB853E",
          "fontFamily": "Klepon Ijo",
          "fontSize": "2.5500000000000003rem",
          "letterSpacing": "0",
        }
      }
    >
      26$
    </p>
  </div>
  <button
    className="sc-bcXHqe sc-pyfCe feNVH hIOFrL icon-button btn-lg button"
    onClick={[Function]}
    style={
      Object {
        "fontSize": "1.05rem",
        "padding": "0.65rem 1.25rem",
      }
    }
  >
    <img
      height={14}
      src={
        Object {
          "ReactComponent": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
          "__esModule": true,
          "default": "trash.svg",
        }
      }
    />
  </button>
</li>
`;
